HeterogeneousCore/CUDAUtilities/src/cudaMemoryPool.cc

```diff
@@ -0,0 +1,148 @@
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaMemoryPool.h"
+
+#include "HeterogeneousCore/CUDAUtilities/interface/SimplePoolAllocator.h"
+
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <cuda_runtime_api.h>
+
+#include <iostream>
+
+namespace {
+
+  //  free callback
+  void CUDART_CB freeCallback(cudaStream_t streamId, cudaError_t status, void *p) {
+    //void CUDART_CB freeCallback(void *p) {
+    if (status != cudaSuccess) {
+      std::cout << "Error in free callaback in stream " << streamId << std::endl;
+      auto error = cudaGetErrorName(status);
+      auto message = cudaGetErrorString(status);
+      std::cout << " error " << error << ": " << message << std::endl;
+    }
+    // std::cout << "free callaback for stream " << streamId << std::endl;
+    auto payload = (memoryPool::Payload *)(p);
+    poolDetails::freeAsync(payload);
+  }
+
+}  // namespace
+
+struct CudaAlloc {
+  static void scheduleFree(memoryPool::Payload *payload, void *stream) {
+    // std::cout    << "schedule free for stream " <<  stream <<std::endl;
+    if (!stream)
+      std::cout << "???? schedule free for stream " << stream << std::endl;
+    if (cudaStreamQuery((cudaStream_t)(stream)) == cudaSuccess)
+      poolDetails::freeAsync(payload);
+    else
+      cudaCheck(cudaStreamAddCallback((cudaStream_t)(stream), freeCallback, payload, 0));
+    // cudaCheck(cudaLaunchHostFunc(stream, freeCallback, payload));
+  }
+};
+
+struct CudaDeviceAlloc : public CudaAlloc {
+  using Pointer = void *;
+
+  static Pointer alloc(size_t size) {
+    Pointer p = nullptr;
+    auto err = cudaMalloc(&p, size);
+    // std::cout << "alloc " << size << ((err == cudaSuccess) ? " ok" : " err") << std::endl;
+    return err == cudaSuccess ? p : nullptr;
+  }
+  static void free(Pointer ptr) {
+    if (ptr == nullptr)
+      std::cout << "free nullptr???" << std::endl;
+    auto err = cudaFree(ptr);
+    // std::cout << "free" << ((err == cudaSuccess) ? " ok" : " err") <<std::endl;
+    if (err != cudaSuccess)
+      std::cout << " error in cudaFree??" << std::endl;
+  }
+};
+
+struct CudaHostAlloc : public CudaAlloc {
+  using Pointer = void *;
+
+  static Pointer alloc(size_t size) {
+    Pointer p = nullptr;
+    auto err = cudaMallocHost(&p, size);
+    // std::cout << "alloc H " << size << ((err == cudaSuccess) ? " ok" : " err") << std::endl;
+    return err == cudaSuccess ? p : nullptr;
+  }
+  static void free(Pointer ptr) { cudaFreeHost(ptr); }
+};
+
+namespace {
+
+  // FIXME : move it in its own place
+  std::unique_ptr<SimplePoolAllocatorImpl<PosixAlloc>> cpuPool;
```

Comment:
I moved to explicit init and shutdown in the CUDAService because if the service is deleted first No cuda "exists" anymore (nice crash in ```garbageCollect()```) and one has to leak memory at the end of the job...
CPU pool follows for symmetry.